Add utility to use `HabanaDataLoader` (#13995)
Lightning-AI/lightning: docs/source-pytorch/accelerators/hpu_intermediate.rst

    trainer = Trainer(accelerator="hpu", callbacks=[device_stats])

For more details, please refer to `Memory Stats APIs <https://docs.habana.ai/en/v1.5.0/PyTorch/PyTorch_User_Guide/Python_Packages.html#memory-stats-apis>`__.

----

> Using HPU Data Module
> ----------------------

Reviewer: ```suggestion
Using HPU Data Module
------------------------
```

> Using HPU Data Module
> ----------------------

Reviewer: ```suggestion
Using HPUDataModule
-----------------------
```

> HPUDataModule class is a wrapper around the pl.LightningDataModule class. It makes working with custom models easier on HPU devices.
> It uses HabanaDataloader for training, test and validation of user-provided models. Currently, it only supports Imagenet dataset.

Reviewer: ```suggestion
``HPUDataModule`` class is a wrapper around the ``LightningDataModule`` class. It makes working with custom models easier on HPU devices.
It uses HabanaDataloader for training, testing, and validation of user-provided models. Currently, it only supports the ``Imagenet`` dataset.
```

> The below snippet shows an example of how to use HPUDataModule.

Reviewer: ```suggestion
The below snippet shows an example of how to use the ``HPUDataModule``.
```

> The below snippet shows an example of how to use the ``HPUDataModule``.

Reviewer 2: ```suggestion
Here's an example of how to use the ``HPUDataModule``:
```

.. code-block:: python

    import pytorch_lightning as pl
    from pytorch_lightning.utilities.hpu_datamodule import HPUDataModule

    train_dir = "./path/to/train/data"
    val_dir = "./path/to/val/data"

    normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])

    train_transforms = [
        transforms.RandomResizedCrop(224),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        normalize,
    ]
    val_transforms = [
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        normalize,
    ]

    data_module = HPUDataModule(
        train_dir,
        val_dir,
        train_transforms=train_transforms,
        val_transforms=val_transforms,
        num_workers=8,
        batch_size=32,
        shuffle=False,
        pin_memory=True,
        drop_last=True,
    )

    # Initialize a trainer
    trainer = pl.Trainer(devices=1, accelerator="hpu", max_epochs=1, max_steps=2)

    # Init our model
    model = RN50Module()  # Or any other model to be defined by user

    trainer.fit(model, datamodule=data_module)
    trainer.validate(model, datamodule=data_module)

> A working example can be found at examples/pl_hpu/hpu_datamodule_sample.py

Reviewer 2: ```suggestion
A working example can be found at ``examples/pl_hpu/hpu_datamodule_sample.py``.
```

> For more details please refer to `Habana dataloader <https://docs.habana.ai/en/latest/PyTorch_User_Guide/PyTorch_User_Guide.html#habana-data-loader>`__.

Reviewer 2: ```suggestion
For more details refer to `Habana dataloader <https://docs.habana.ai/en/latest/PyTorch_User_Guide/PyTorch_User_Guide.html#habana-data-loader>`__.
```
